query-algebrizer/src/types.rs

```diff
@@ -85,4 +90,117 @@ impl QualifiedAlias {
     pub fn for_type_tag(&self) -> QualifiedAlias {
         QualifiedAlias(self.0.clone(), DatomsColumn::ValueTypeTag)
     }
-}
\ No newline at end of file
+}
+
+#[derive(PartialEq, Eq)]
+pub enum QueryValue {
+    Column(QualifiedAlias),
+    Entid(Entid),
+    TypedValue(TypedValue),
+
+    // This is different: a numeric value can only apply to the 'v' column, and it implicitly
+    // constrains the `value_type_tag` column. For instance, a primitive long on `datoms00` of `5`
+    // cannot be a boolean, so `datoms00.value_type_tag` must be in the set `#{0, 4, 5}`.
+    // Note that `5 = 5.0` in SQLite, and we preserve that here.
+    PrimitiveLong(i64),
+}
+
+impl Debug for QueryValue {
+    fn fmt(&self, f: &mut Formatter) -> ::std::fmt::Result {
+        use self::QueryValue::*;
+        match self {
+            &Column(ref qa) => {
+                write!(f, "{:?}", qa)
+            },
+            &Entid(ref entid) => {
+                write!(f, "entity({:?})", entid)
+            },
+            &TypedValue(ref typed_value) => {
+                write!(f, "value({:?})", typed_value)
+            },
+            &PrimitiveLong(value) => {
+                write!(f, "primitive({:?})", value)
+            },
+
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+/// Define the different numeric inequality operators that we support.
+/// Note that we deliberately don't just use "<=" and friends as strings:
+/// Datalog and SQL don't use the same operators (e.g., `<>` and `!=`).
+pub enum NumericComparison {
+    LessThan,
+    LessThanOrEquals,
+    GreaterThan,
+    GreaterThanOrEquals,
+    NotEquals,
+}
+
+impl NumericComparison {
+    pub fn to_sql_operator(self) -> &'static str {
+        use self::NumericComparison::*;
+        match self {
+            LessThan            => "<",
+            LessThanOrEquals    => "<=",
+            GreaterThan         => ">",
+            GreaterThanOrEquals => ">=",
+            NotEquals           => "<>",
+        }
+    }
+
+    pub fn from_datalog_operator(s: &str) -> Option<NumericComparison> {
+        match s {
+            "<"  => Some(NumericComparison::LessThan),
+            "<=" => Some(NumericComparison::LessThanOrEquals),
+            ">"  => Some(NumericComparison::GreaterThan),
+            ">=" => Some(NumericComparison::GreaterThanOrEquals),
+            "!=" => Some(NumericComparison::NotEquals),
+            _    => None,
+        }
+    }
+}
+
+impl Debug for NumericComparison {
+    fn fmt(&self, f: &mut Formatter) -> ::std::fmt::Result {
+        use self::NumericComparison::*;
+        f.write_str(match self {
+            &LessThan => "<",
+            &LessThanOrEquals => "<=",
+            &GreaterThan => ">",
+            &GreaterThanOrEquals => ">=",
+            &NotEquals => "!=",                // Datalog uses !=. SQL uses <>.
```

Comment:
It does; one of SQLite's design goals is to do the best it can with SQL written for other database systems. But they all handle `<>`, so that's what we'll use.